Sources/AsyncHTTPClient/ConnectionPool/ChannelHandler/HTTP1ProxyConnectHandler.swift

```diff
@@ -0,0 +1,204 @@
+                // Any 2xx (Successful) response indicates that the sender (and all
+                // inbound proxies) will switch to tunnel mode immediately after the
+                // blank line that concludes the successful response's header section
+                self.state = .headReceived(promise)
+            case 407:
+                let error = HTTPClientError.proxyAuthenticationRequired
+                self.state = .failed(error)
+                self.proxyEstablishedPromise?.fail(error)
+                context.fireErrorCaught(error)
+                context.close(mode: .all, promise: nil)
+
+            default:
+                // Any response other than a successful response
+                // indicates that the tunnel has not yet been formed and that the
+                // connection remains governed by HTTP.
+                let error = HTTPClientError.invalidProxyResponse
+                self.state = .failed(error)
+                self.proxyEstablishedPromise?.fail(error)
+                context.fireErrorCaught(error)
+                context.close(mode: .all, promise: nil)
+            }
+        case .body:
+            switch self.state {
+            case .headReceived(let timeout):
+                timeout.cancel()
+                // we don't expect a body
+                let error = HTTPClientError.invalidProxyResponse
+                self.state = .failed(error)
+                self.proxyEstablishedPromise?.fail(error)
+                context.fireErrorCaught(error)
+                context.close(mode: .all, promise: nil)
+
+            case .failed:
+                // ran into an error before... ignore this one
+                break
+            case .completed, .connectSent, .initialized:
+                preconditionFailure("Invalid state: \(self.state)")
+            }
+
+        case .end:
+            switch self.state {
+            case .headReceived(let timeout):
+                timeout.cancel()
+                self.state = .completed
+                self.proxyEstablishedPromise?.succeed(())
+
+            case .failed:
+                // ran into an error before... ignore this one
+                break
+            case .initialized, .connectSent, .completed:
+                preconditionFailure("Invalid state: \(self.state)")
+            }
+        }
+    }
+
+    func sendConnect(context: ChannelHandlerContext) {
+        guard case .initialized = self.state else {
+            // we might run into this handler twice, once in handlerAdded and once in channelActive.
+            return
+        }
+
+        let timeout = context.eventLoop.scheduleTask(deadline: self.deadline) {
+            switch self.state {
+            case .initialized:
+                preconditionFailure("How can we have a scheduled timeout, if the connection is not even up?")
+
+            case .connectSent(let scheduled), .headReceived(let scheduled):
+                scheduled.cancel()
```

Comment:
We can omit the call to `cancel` here, we know the scheduled task fired if we're in this function!